// endskin/endskin.js
var RootPath = '';
var fs = require('fs');
var S = require('string');

module.exports.setRoot = function(p)//¼ì²é¸ùÄ¿Â¼Â·¾¶ÊÇ·ñ×îºó¼Ó'/'
{
	RootPath = p;
	if (!RootPath.match(/\/$/)) RootPath+='/';
};

module.exports.create = function(f)
{
	return new EndSkin(f);
};

exports.cache = {};

function EndSkin(tmpId)
{
	if (!tmpId)
	{
		throw new Error('EndSkin: no template name passed');
		return;
	}
	var codeBlocks = [];

	this.templateReader = function(f,cb)
	{
		var fMatch = f.match(/^(.*?)\w/i);
		f = f.replace(fMatch[1],'');
		var file = RootPath+f;
		// var fileMatch = file.match(/views\/(.*?)\w/i);
		// file = file.replace(fileMatch[1],'');
		// console.log(file);
		return fs.readFileSync(file,'utf-8');
	} 
	
	this.getTemplateString = function(id)
	{
		var s = this.templateReader(id);
		if (!s)
		{
			throw new Error('EndSkin: template "'+id+'" not found');
			return;
		}
		//TRIP.log(id);
		//TRIP.log(s);


		var ms;
		// 删除jsp注释
		ms = s.match(/<%--(.|\n|\r)*?--%>/ig);
		if (ms) {
			for(i = 0; m = ms[i]; i++){
				s = s.replace(m,'');
			}
		}

		// 删除<!-- -->注释
		ms = s.match(/<!--(.|\n|\r)*?-->/ig);
		if (ms) {
			for(i = 0; m = ms[i]; i++){
				s = s.replace(m,'');
			}
		}

		// 防止taglib.jsp|namespace.jsp的引入
		ms =s.match(/<%@.*?include.*file=\".*?(taglib.jsp|namespace.jsp)\".*?%>/ig);
		if (ms) {
			for(i = 0; m = ms[i]; i++){
				s = s.replace(m,'');
			}
		}

		// 移除<%@ page ... %>
		ms = s.match(/<%@.*?page.*?%>/ig);
		if (ms) {
			for(i = 0; m = ms[i]; i++){
				s = s.replace(m,'');
			}
		}

		// 读取include 中的jsp，并将其导入
		ms = s.match(/<%@.*?include.*file=\"(.*?)\".*?%>/gi);
		if (ms){
			 for(var i=0,m;m=ms[i]; i++)
				{
					var _ms = m.match(/<%@.*?include.*file=\"(.*?)\".*?%>/i);
					s = s.replace(m,this.getTemplateString(_ms[1]));
				}
		}

		return s;
	}
	
	this.data = {};
	
	this.assign = function(key,val)
	{
		if (key && val) this.data[key] = val;
		else if (typeof key == 'object')
			for(var k in key)
				this.data[k] = key[k];
		return this;
	}
	
	
	this.compile = function(page)
	{
		var ms = null,
			i=0,
			m;
		var _ms = null,
			item,
			variable,
			variableList = [],
			value,
			cond,
			tagCount = {
				setBegin : 0,
				setEnd : 0,
				forEachBegin : 0,
				forEachEnd : 0,
				ifBegin : 0,
				ifEnd : 0,
				whenBegin : 0,
				whenEnd : 0,
				chooseBegin : 0,
				chooseEnd : 0,
				otherwiseBegin : 0,
				otherwiseEnd : 0
			};

		if (ms = page.match(/\<c\:set.*\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				_ms = m.match(/\<c\:set.*?var=\"(.*?)\".*?value=\"(\$\{.*?\}).*?\"\>/i);
				if (_ms)
				{
					value = _ms[2];
					variable = '${'+_ms[1]+'}';
					page = S(page).replaceAll(variable,value).s;
					tagCount['setBegin']++;
				}
				page = page.replace(m,'');
			}
		}
		
		// 匹配</c:set>并替换为空
		if (ms = page.match(/\<\/c\:set.*\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				page = page.replace(m,'');
				tagCount['setEnd']++;
			}
		}
		
		// 匹配<c:forEach>并使用for in处理
		if (ms = page.match(/\<c\:forEach.+\$\{(.*)\}.*var=\"([a-zA-Z0-9]+)\".*?\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
//				console.log("m:" + m);
				_ms = m.match(/\<c\:forEach.+\$\{(.*)\}.*var=\"([a-zA-Z0-9]+)\".*?\>/i);
				if (_ms)
				{
					variable = this._replace_var_name_jsp(_ms[1]);
					item = _ms[2];
					code = '(function()\n{\nvar __var = '+variable+';\nfor(var __key in __var)\n{\nthis.data[\"'+item+'\"] = __var[__key]; \n';
					codeBlocks.push(code);
					page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
					continue;
				}
				tagCount['forEachBegin']++;
			}
		}
		
		if (ms = page.match(/\<\/c:forEach>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				code = '}\n}).call(this);\n';
				codeBlocks.push(code);
				page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				tagCount['forEachEnd']++;
			}
		}

		// 匹配形如<c:if|when test="${not empty y}">的标签
		if (ms = page.match(/\<c\:(if|when).*?test=\"\$\{\s*not\s*empty\s*(.*?)\s*\}.*?\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				_ms = m.match(/\<c\:(if|when).*?test=\"\$\{\s*not\s*empty\s*(.*?)\s*\}.*?\>/i);
				if (_ms)
				{
					variable = this._replace_var_name_jsp(_ms[2]);
					if (_ms[1] === 'if') {
						code = 'if ('+ variable + '!= "")\n{\n';
					}else if(_ms[1] === 'when'){
						code = 'case  ('+ variable + '!= "") : ';
					}		
					tagCount[_ms[1]+'Begin']++;		
					codeBlocks.push(code);
					page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				}
			}
		}

		// 匹配形如<c:if|when test="${empty y}">的标签
		if (ms = page.match(/\<c\:(if|when).*?test=\"\$\{\s*empty\s*(.*?)\s*\}.*?\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				_ms = m.match(/\<c\:(if|when).*?test=\"\$\{\s*empty\s*(.*?)\s*\}.*?\>/i);
				if (_ms)
				{
					variable = this._replace_var_name_jsp(_ms[2]);
					if (_ms[1] === 'if') {
						code = 'if ('+ variable + '== "")\n{\n';
					}else if(_ms[1] === 'when'){
						code = 'case  ('+ variable + '== "") : ';
					}
					tagCount[_ms[1]+'Begin']++;
					codeBlocks.push(code);
					page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				}
			}
		}

		// 匹配形如<c:if|when test="${y eq x}">的标签
		if (ms = page.match(/\<c\:(if|when).*?test=\"\$\{\s*(.*?)\s+eq\s+(.*?)\s*\}.*?\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				_ms = m.match(/\<c\:(if|when).*?test=\"\$\{\s*(.*?)\s+eq\s+(.*?)\s*\}.*?\>/i);
				if (_ms)
				{
					variableList[0] = this._replace_var_name_jsp(_ms[2]);
					variableList[1] = this._replace_var_name_jsp(_ms[3]);
					if (_ms[1] === 'if') {
						code = 'if ('+ variableList[0] +'=='+ variableList[1]+')\n{\n';
					}else if(_ms[1] === 'when'){
						code = 'case ('+ variableList[0] +'=='+ variableList[1]+') : ';
					}
					tagCount[_ms[1]+'Begin']++;
					codeBlocks.push(code);
					page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
					// console.log('***'+_ms[1]+'***'+_ms[2]+'***'+_ms[3])
					// console.log(code);
				}
			}
		}

		// 匹配形如<c:if|when test="${y > 3}">的标签
		if (ms = page.match(/\<c\:(if|when).*?test=\"\$\{\s*(.*?)([\>\<\=].*?)\s*\}.*?\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				_ms = m.match(/\<c\:(if|when).*?test=\"\$\{\s*(.*?)([\>\<\=].*?)\s*\}.*?\>/i);
				if (_ms)
				{
					variable = this._replace_var_name_jsp(_ms[2]);
					cond = _ms[3];
					if (_ms[1] === 'if') {
						code = 'if ('+ variable + cond+')\n{\n';
					}else if(_ms[1] === 'when'){
						code = 'case ('+ variable + cond+') :';
					}
					tagCount[_ms[1]+'Begin'];
					codeBlocks.push(code);
					page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
					// console.log('***'+_ms[1]+'***'+_ms[2]+'***'+_ms[3])
				}
			}
		}
		//  匹配形如<c:if|when test=${y}>的标签
		if (ms = page.match(/\<c\:(if|when).*?test=\"\$\{\s*(.*?)\s*\}.*?\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				_ms = m.match(/\<c\:(if|when).*?test=\"\$\{\s*(.*?)\s*\}.*?\>/i);
				if (_ms)
				{
					variable = this._replace_var_name_jsp(_ms[2]);
					if (_ms[1] === 'if') {
						code = 'if ('+ variable +' == true)\n{\n';
					}else if(_ms[1] === 'when'){
						code = 'case ('+ variable +' == true) : ';
					}				
					tagCount[_ms[1]+'Begin'];	
					codeBlocks.push(code);
					page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				}
			}
		}
		
		//  匹配形如<\c:if>的标签
		if (ms = page.match(/\<\/c\:if>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				code = '}\n';
				codeBlocks.push(code);
				page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				tagCount['ifEnd']++
			}
		}

		// 匹配</c:when>标签
		if (ms = page.match(/\<\/c\:when>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				code = '\nbreak;\n';
				codeBlocks.push(code);
				page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				tagCount['whenEnd']++;
			}
		}
		
		// 匹配<c:otherwise>替换为switch语句的 default
		if (ms = page.match(/\<c\:otherwise\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				code = 'default :';
				codeBlocks.push(code);
				page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				tagCount['otherwiseBegin']++;
			}
		}
		
		if (ms = page.match(/\<\/c\:otherwise\>/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				code = '\nbreak;';
				codeBlocks.push(code);
				page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				tagCount['otherwiseEnd']++;
			}
		}

		// 匹配<c:choose>和</c:choose>替换switch (true){}
		if(ms = page.match(/\<c\:choose>/ig)){
			for(i=0; m=ms[i]; i++){
				code = '\nswitch (true) {\n';
				codeBlocks.push(code);
				page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				tagCount['chooseBegin']++;
			}
		}
		
		if(ms = page.match(/\<\/c\:choose>/ig)){
			for(i=0; m=ms[i]; i++){
				code = '\n}\n';
				codeBlocks.push(code);
				page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
				tagCount['chooseEnd']++;
			}
		}
		
		// 匹配${variable}并替换为data中的数据
		if (ms = page.match(/\$\{([a-zA-Z\_][a-zA-Z0-9\_\.\[\]\'\"]*).*?\}/ig))
		{
			for(i=0; m=ms[i]; i++)
			{
				_ms = m.match(/\$\{([a-zA-Z\_][a-zA-Z0-9\_\.\[\]\'\"]*).*?\}/i);
				code = 'output.push('+this._replace_var_name_jsp(_ms[1])+');';
				codeBlocks.push(code);
				page = page.replace(m,'{{{{EndSkin.codeblock['+(codeBlocks.length-1)+']}}}}');
			}
		}
		
		var arr = page.split(/\{\{\{\{EndSkin\.codeblock\[\d+\]\}\}\}\}/);
		var ms = page.match(/\{\{\{\{EndSkin\.codeblock\[(\d+)\]\}\}\}\}/g);

		var codes = ['var output = [];'];
		for(i=0;i<arr.length-1; i++)
		{
				if(arr[i].replace(/\s+/g,"") != ""){
					codes.push('output.push('+JSON.stringify(arr[i])+');\n');
				}
				var _ms = ms[i].match(/\{\{\{\{EndSkin\.codeblock\[(\d+)\]\}\}\}\}/);
				codes.push(codeBlocks[parseInt(_ms[1])]+'\n');
			
		}
		codes.push('output.push('+JSON.stringify(arr.pop())+');\n');
		codes.push("return output.join('');");
		
		// console.log("codes:" + codes.join(''));

		try
		{
			return new Function(codes.join(''));
		}
		catch(e)
		{
			var err = [];
			if (tagCount['forEachBegin'] > tagCount['forEachEnd']) err.push('missing '+(tagCount['forEachBegin'] - tagCount['forEachEnd'])+' {/foreach}');
			if (tagCount['forEachBegin'] < tagCount['forEachEnd']) err.push('too much {/foreach}');
//			if (cOthersBegins != foreachEnds) err.push('too much {/foreach}');
			if (tagCount['ifBegin'] > tagCount['ifEnd']) err.push('missing '+(tagCount['ifBegin'] - tagCount['ifEnd'])+' {/if}');
			if (tagCount['ifBegin'] < tagCount['ifEnd']) err.push('too much {/if}');
			if (tagCount['chooseBegin'] > tagCount['chooseEnd']) err.push('missing '+(tagCount['chooseBegin'] - tagCount['chooseEnd'])+' {/choose}');
			if (tagCount['chooseBegin'] < tagCount['chooseEnd']) err.push('too much {/choose}');
			if (tagCount['whenBegin'] > tagCount['whenEnd']) err.push('missing '+(tagCount['whenBegin'] - tagCount['whenEnd'])+' {/when}');
			if (tagCount['whenBegin'] < tagCount['whenEnd']) err.push('too much {/when}');
			if (tagCount['otherwiseBegin'] > tagCount['otherwiseEnd']) err.push('missing '+(tagCount['otherwiseBegin'] - tagCount['otherwiseEnd'])+' {/otherwise}');
			if (tagCount['otherwiseBegin'] < tagCount['otherwiseEnd']) err.push('too much {/otherwise}');
			var re = 'EndSkin Compile Error: \nview file:'+tmpId+'\n'+e.toString() + ';\n' + err.join(';\n');
			throw new Error(re);
			return re;
		}
	}
	
	this._replace_var_name_jsp = function(s)
	{
//		console.log("_replace_var_name_jsp:" + s);
		s = s.replace(/[a-z\_0-9\.]+/ig,function(s)
		{
			return 'this.show_val_jsp(\"' + s + '\")';
		});
		return s;
	}
	
	
	
	this.show_val_jsp = function(s)
	{
		var arr = s.split('.');
		var val = this.data;
		for(var i=0;i<arr.length;i++)
		{
			var key = arr[i];
			if (val[key] == undefined) return '';
			val = val[key];
		}
		if (typeof val == 'function') val = val();
		return val;
	}
	
	this._replace_var_name = function(s)
	{
		s = s.replace(/\$[a-z\_0-9\.]+/ig,function(s)
		{
			return 'this.show_val(\"' + s.replace('$','') + '\")';
		});
		return s;
	}
	
	
	
	this.show_val = function(s)
	{
		var arr = s.split('.');
		var val = this.data;
		for(var i=0;i<arr.length;i++)
		{
			var key = arr[i];
			if (val[key] == undefined) return '';
			val = val[key];
		}
		if (typeof val == 'function') val = val();
		return val;
	}
	
	var s = this.getTemplateString(tmpId);
	//console.log(s);
	if (!s) return;
	var cached_func = this.compile(s);
	
	this.html = function()
	{
		if (typeof cached_func == 'function')
			return cached_func.call(this);
		else
			return cached_func.toString();
	};
//	console.log(this.html());
	
	return this;
};



/*
* express 3.x support
*/
exports.renderFile = function(path, options, fn)
{
	console.log(path)
	var key = path + ':func';

	// console.log(path);
	if('function' == typeof options) 
	{
		fn = options, options = {};
	}

	if (options.settings && options.settings.views)
	{
		RootPath = options.settings.views;
	}
	else
	{
		RootPath = path.replace(/\/[^\/]*$/,'/');
	}
	
	var viewsMatch = path.match(/(views.*?)\w/i);

	var tmpId = path.split(viewsMatch[1]).pop();
	try 
	{
		if (options.cache && exports.cache[key])
		{
			exports.cache[key].data = options;
			fn(null,exports.cache[key].html());
		}
		else
		{
			var func = new EndSkin(tmpId);
			exports.cache[key] = func;
			func.assign(options);
			fn(null,func.html());
		}
	}
	catch(err) 
	{
		fn(err);
	}
};


exports.__express = exports.renderFile;
